### PR TITLE
userguide of chimera

### PR DESCRIPTION
**Pull request by JkSelf**
Complete the userguide of chimera.

### CODE REVIEW

**Comment by sundapeng** [comment]
Hi @JkSelf , could you update your branch gh-pages with your target/site, so that people can review the contents directly.

**Comment by JkSelf** [comment]
Hi @sundapeng  I have update the gh-pages branch with target/site. And the link is http://jkself.github.io/chimera/userguide.html.

File: src/site/xdoc/userguide.xml [hunk 1]
```diff
@@ -25,8 +25,98 @@
         algorithm such as AES directly. It wraps to Openssl or JCE which implement the algorithms.
       </p>
       <subsection name="Example Title">
```
Comment (dianfu) [inline src/site/xdoc/userguide.xml]: "Example Title" can be changed to something like "Interfaces Overview"

File: src/site/xdoc/userguide.xml [hunk 2]
```diff
@@ -25,8 +25,98 @@
...
-        <p>TODO</p>
+        <p>Interfaces and classes used by the various implementation in the sub-packages.</p>
+        <table>
+          <tr>
+            <td width="150">
+              <a href="http://project.sundp.me/chimera/apidocs/index.html">cipher
```
Comment (dianfu) [inline src/site/xdoc/userguide.xml]: The hyperlink is wrong.

File: src/site/xdoc/userguide.xml [hunk 3]
```diff
@@ -25,8 +25,98 @@
...
+              </a>
+            </td>
+            <td>
+              The interface of cryptographic cipher for encryption and decryption.
+            </td>
+          </tr>
+
+          <tr>
+            <td width="150">
+              <a href="http://project.sundp.me/chimera/apidocs/index.html">random
```
Comment (dianfu) [inline src/site/xdoc/userguide.xml]: The hyperlink is wrong.

File: src/site/xdoc/userguide.xml [hunk 4]
```diff
@@ -25,8 +25,98 @@
...
+              </a>
+            </td>
+            <td>
+              The interface for SecureRandom.
+            </td>
+          </tr>
+
+          <tr>
+            <td width="150">
+              <a href="http://project.sundp.me/chimera/apidocs/index.html">stream
```
Comment (dianfu) [inline src/site/xdoc/userguide.xml]: The hyperlink is wrong.

File: src/site/xdoc/userguide.xml [hunk 5]
```diff
@@ -25,8 +25,98 @@
...
+              </a>
+            </td>
+            <td>
+              The Input/Output interface abstract the input/output source of CryptoInputStream/CryptoOutputStream so that
```
Comment (dianfu) [inline src/site/xdoc/userguide.xml]: The interface wraps the underlying stream and it automatically encrypts the stream when data is written and decrypts the stream when data is read.

File: src/site/xdoc/userguide.xml [hunk 6]
```diff
@@ -25,8 +25,98 @@
...
+              <h4>Usage of Cipher API</h4>
+              Properties properties = new Properties();<br/>
+              // Default value is com.intel.chimera.crypto.JceCipher<br/>
+              properties.setProperty("chimera.crypto.cipher.classes", "com.intel.chimera.crypto.OpensslCipher,com.intel.chimera.crypto.JceCipher");<br/>
+              Cipher cipher = Utils.getCipherInstance(CipherTransformation.AES_CTR_NOPADDING, properties);<br/><br/>
+              // Available key length is: 16, 24, 32 bytes<br/>
+              byte[] key = new byte[16];<br/>
+              // IV length is always 16 bytes<br/>
+              byte[] iv = new byte[16];<br/>
+              // 1 represents encryption ;0 represents decryption<br/>
```
Comment (dianfu) [inline src/site/xdoc/userguide.xml]: minor typo issue: "encryption ;0 represents" should be "encryption; 0 represents"

File: src/site/xdoc/userguide.xml [hunk 7]
```diff
@@ -24,9 +24,98 @@
         the minimum code and effort. Please note that Chimera doesn't implement the cryptographic
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Chimera -> Apache Commons Crypto

File: src/site/xdoc/userguide.xml [hunk 8]
```diff
@@ -24,9 +24,98 @@
...
+
+        </table>
       </subsection>
+      <subsection name ="Usage">
+        <table>
+          <tr>
+            <td>
+              <ol style="list-style-type: decimal">
+              <h4>Usage of Random API</h4>
+              byte[] data = new byte[16];<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Better to add comments: //Construct a byte array to store random data.

File: src/site/xdoc/userguide.xml [hunk 9]
```diff
@@ -24,9 +24,98 @@
...
+              Properties properties = new Properties();<br/>
+              // Default value is com.intel.chimera.random.JavaSecureRandom<br/>
+              properties.setProperty("chimera.crypto.secure.random.classes", "com.intel.chimera.random.OpensslSecureRandom,com.intel.chimera.random.OsSecureRandom");<br/>
+              SecureRandomFactory.getSecureRandom(properties).nextBytes(data);<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Better to separate to two lines.
````
// Gets the `SecureRandom` instance.
SecureRandom secureRandom = SecureRandomFactory.getSecureRandom(properties)
// Generates random bytes and places them into the byte array.
secureRandom.nextBytes(data);

File: src/site/xdoc/userguide.xml [hunk 10]
```diff
@@ -24,9 +24,98 @@
...
+              <h4>Usage of Cipher API</h4>
+              Properties properties = new Properties();<br/>
+              // Default value is com.intel.chimera.crypto.JceCipher<br/>
+              properties.setProperty("chimera.crypto.cipher.classes", "com.intel.chimera.crypto.OpensslCipher,com.intel.chimera.crypto.JceCipher");<br/>
+              Cipher cipher = Utils.getCipherInstance(CipherTransformation.AES_CTR_NOPADDING, properties);<br/><br/>
+              // Available key length is: 16, 24, 32 bytes<br/>
+              byte[] key = new byte[16];<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: new byte[16] will be the empty byte [0,0,..0], we are better to add an example byte array with values, and add comments likes:
````
// Constructs the crypto key for cipher.

File: src/site/xdoc/userguide.xml [hunk 11]
```diff
@@ -24,9 +24,98 @@
...
+              // IV length is always 16 bytes<br/>
+              byte[] iv = new byte[16];<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Same as the comments for `key`: new byte[16] will be the empty byte [0,0,..0], better to add the code and comments which are using SecureRandom generate the IV.

File: src/site/xdoc/userguide.xml [hunk 12]
```diff
@@ -24,9 +24,98 @@
...
+              // 1 represents encryption; 0 represents decryption<br/>
+              int mode = 1;<br/>
+              cipher.init(mode, key, iv);<br/><br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: mode -> `Cipher.ENCRYPT_MODE` or `Cipher.ENCRYPT_MODE`

File: src/site/xdoc/userguide.xml [hunk 13]
```diff
@@ -24,9 +24,98 @@
...
+              <h4>Usage of Cipher API</h4>
+              Properties properties = new Properties();<br/>
+              // Default value is com.intel.chimera.crypto.JceCipher<br/>
+              properties.setProperty("chimera.crypto.cipher.classes", "com.intel.chimera.crypto.OpensslCipher,com.intel.chimera.crypto.JceCipher");<br/>
+              Cipher cipher = Utils.getCipherInstance(CipherTransformation.AES_CTR_NOPADDING, properties);<br/><br/>
+              // Available key length is: 16, 24, 32 bytes<br/>
+              byte[] key = new byte[16];<br/>
+              // IV length is always 16 bytes<br/>
+              byte[] iv = new byte[16];<br/>
+              // 1 represents encryption; 0 represents decryption<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Better to use Cipher.ENCRYPT_MODE, so we don't need the comments.

File: src/site/xdoc/userguide.xml [hunk 14]
```diff
@@ -24,9 +24,98 @@
...
+              int inputLen = input.length();<br/>
+              byte[] output = new byte[1024];<br/>
+              int outputOffset = 0;<br/>
+              cipher.update(input.getBytes("UTF-8"), inputOffset, inputLen, output, outputOffset);<br/><br/>
+              int bufferSize = 4096;<br/>
+              ByteBuffer inBuffer = ByteBuffer.allocateDirect(bufferSize);<br/>
+              ByteBuffer outBuffer = ByteBuffer.allocateDirect(bufferSize);<br/>
+              inBuffer.put("The data you want to encrypt or decrypt".getBytes("UTF-8"));<br/>
+              cipher.update(inBuffer, outBuffer);<br/>
+              cipher.doFinal(inBuffer, outBuffer);<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Add comment: we should call do final at the end of encryption/decryption

File: src/site/xdoc/userguide.xml (hunk 12 above)
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Add comments: Initializes the cipher with mode, key and iv.

File: src/site/xdoc/userguide.xml [hunk 15]
```diff
@@ -24,9 +24,98 @@
...
+              byte[] iv = new byte[16];<br/>
+              // 1 represents encryption; 0 represents decryption<br/>
+              int mode = 1;<br/>
+              cipher.init(mode, key, iv);<br/><br/>
+              String input = "hello world!";<br/>
+              int inputOffset = 0;<br/>
+              int inputLen = input.length();<br/>
+              byte[] output = new byte[1024];<br/>
+              int outputOffset = 0;<br/>
+              cipher.update(input.getBytes("UTF-8"), inputOffset, inputLen, output, outputOffset);<br/><br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Add Comments: 
````
Continues a multiple-part encryption/decryption operation for byte array.

File: src/site/xdoc/userguide.xml [hunk 16]
```diff
@@ -24,9 +24,98 @@
...
+              int bufferSize = 4096;<br/>
+              ByteBuffer inBuffer = ByteBuffer.allocateDirect(bufferSize);<br/>
+              ByteBuffer outBuffer = ByteBuffer.allocateDirect(bufferSize);<br/>
+              inBuffer.put("The data you want to encrypt or decrypt".getBytes("UTF-8"));<br/>
+              cipher.update(inBuffer, outBuffer);<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Add Comments: 
````
Continues a multiple-part encryption/decryption operation for byte buffer.

File: src/site/xdoc/userguide.xml [hunk 17]
```diff
@@ -24,9 +24,98 @@
...
+          <tr>
+            <td>
+              <ol style="list-style-type: decimal">
+              <h4>Usage of Random API</h4>
+              byte[] data = new byte[16];<br/>
+              Properties properties = new Properties();<br/>
+              // Default value is com.intel.chimera.random.JavaSecureRandom<br/>
+              properties.setProperty("chimera.crypto.secure.random.classes", "com.intel.chimera.random.OpensslSecureRandom,com.intel.chimera.random.OsSecureRandom");<br/>
+              SecureRandomFactory.getSecureRandom(properties).nextBytes(data);<br/>
+              <h4>Usage of Cipher API</h4>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Better to separate to two part:
Usage of Byte Array Encryption/Decryption
Usage of ByteBuffer Encryption/Decryption

File: src/site/xdoc/userguide.xml [hunk 18]
```diff
@@ -24,9 +24,98 @@
...
+              int outputOffset = 0;<br/>
+              cipher.update(input.getBytes("UTF-8"), inputOffset, inputLen, output, outputOffset);<br/><br/>
+              int bufferSize = 4096;<br/>
+              ByteBuffer inBuffer = ByteBuffer.allocateDirect(bufferSize);<br/>
+              ByteBuffer outBuffer = ByteBuffer.allocateDirect(bufferSize);<br/>
+              inBuffer.put("The data you want to encrypt or decrypt".getBytes("UTF-8"));<br/>
+              cipher.update(inBuffer, outBuffer);<br/>
+              cipher.doFinal(inBuffer, outBuffer);<br/>
+              <h4>Usage of Stream API</h4>
+              byte[] key = new byte[16];<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Same as the comment for cipher key: new byte[16] will be the empty byte [0,0,..0], we are better to add an example byte array with values, and add comments likes:

File: src/site/xdoc/userguide.xml [hunk 19]
```diff
@@ -24,9 +24,98 @@
...
+              byte[] iv = new byte[16];<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Same as the comments for cipher iv: new byte[16] will be the empty byte [0,0,..0], better to add the code and comments which are using SecureRandom generate the IV.

**Comment by JkSelf** [comment]
Thanks @sundapeng @dianfu comments. I have fixed.

File: src/site/xdoc/userguide.xml [hunk 20]
```diff
@@ -21,12 +21,155 @@
...
+              //Constructs a byte array to store random data.<br/>
+              byte[] data = new byte[16];<br/>
+              Properties properties = new Properties();<br/>
+              // Default value is com.intel.chimera.random.JavaSecureRandom.<br/>
+              properties.setProperty("chimera.crypto.secure.random.classes", "com.intel.chimera.random.OpensslSecureRandom,com.intel.chimera.random.OsSecureRandom");<br/>
+              //Gets the 'SecureRandom' instance.<br/>
+              SecureRandom secureRandom = SecureRandomFactory.getSecureRandom(properties);<br/>
+              //Generates random bytes and places them into the byte array.<br/>
+              secureRandom.nextBytes(data);<br/>
+              <h4>Usage of Byte Array Encryption/Decryption</h4>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: We should add close to SecureRandom 
````
secureRandom.close();

File: src/site/xdoc/userguide.xml [hunk 21]
```diff
@@ -21,12 +21,155 @@
...
+              // Available key length is: 16, 24, 32 bytes.<br/>
+              //Constructs the crypto key for cipher.<br/>
+              byte[] key = new byte[16];<br/>
+              //Generates random bytes and places them into the byte array.<br/>
+              secureRandom.nextBytes(key);<br/>
+              // IV length is always 16 bytes.<br/>
+              //Constructs the crypto iv for cipher.<br/>
+              byte[] iv = new byte[16];<br/>
+              //Generates random bytes and places them into the byte array.<br/>
+              secureRandom.nextBytes(iv);<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Also add secureRandom.close(); for cipher

File: src/site/xdoc/userguide.xml [hunk 22]
```diff
@@ -21,12 +21,155 @@
...
+              String input = "hello world!";<br/>
+              int inputOffset = 0;<br/>
+              int inputLen = input.length();<br/>
+              byte[] output = new byte[1024];<br/>
+              int outputOffset = 0;<br/>
+              //Continues a multiple-part encryption/decryption operation for byte array.<br/>
+              cipher.update(input.getBytes("UTF-8"), inputOffset, inputLen, output, outputOffset);<br/>
+              //We should call do final at the end of encryption/decryption.<br/>
+              cipher.doFinal(inBuffer, outBuffer);<br/>
+              <h4>Usage of ByteBuffer Encryption/Decryption</h4>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Can we make `Usage of ByteBuffer Encryption/Decryption` and `Usage of Byte Array Encryption/Decryption` as subtasks for Usage of Cipher API ?

File: src/site/xdoc/userguide.xml [hunk 23]
```diff
@@ -21,12 +21,155 @@
...
+
+        </table>
       </subsection>
+      <subsection name ="Usage">
+        <table>
+          <tr>
+            <td>
+              <ol style="list-style-type: decimal">
+              <h4>Usage of Random API</h4>
+              //Constructs a byte array to store random data.<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Nit: Can we only use the css with background for the content of `Usage of XX`

File: src/site/xdoc/userguide.xml [hunk 24]
```diff
@@ -21,12 +21,155 @@
...
+              // IV length is always 16 bytes<br/>
+              //Constructs the crypto iv for cipher.<br/>
+              byte[] iv = new byte[16];<br/>
+              //Generates random bytes and places them into the byte array.<br/>
+              secureRandom.nextBytes(iv);<br/>
+              int bufferSize = 4096;<br/>
+              String input = "hello world!";<br/>
+              byte[] decryptedData = new byte[1024];<br/>
+              // Encrypt<br/>
+              ByteArrayOutputStream os = new ByteArrayOutputStream();<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: // Constructs the original OutputStream

File: src/site/xdoc/userguide.xml [hunk 25]
```diff
@@ -21,12 +21,155 @@
...
+              secureRandom.nextBytes(key);<br/>
+              // IV length is always 16 bytes<br/>
+              //Constructs the crypto iv for cipher.<br/>
+              byte[] iv = new byte[16];<br/>
+              //Generates random bytes and places them into the byte array.<br/>
+              secureRandom.nextBytes(iv);<br/>
+              int bufferSize = 4096;<br/>
+              String input = "hello world!";<br/>
+              byte[] decryptedData = new byte[1024];<br/>
+              // Encrypt<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: Encryption with `CryptoOutputStream`

File: src/site/xdoc/userguide.xml [hunk 26]
```diff
@@ -21,12 +21,155 @@
...
+              ByteArrayOutputStream os = new ByteArrayOutputStream();<br/>
+              CryptoOutputStream cos = new CryptoOutputStream(os, cipher, bufferSize, key, iv);<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: // Constructs the instanse of CryptoOutputStream.

File: src/site/xdoc/userguide.xml [hunk 27]
```diff
@@ -21,12 +21,155 @@
...
+              cos.write(input.getBytes("UTF-8"));<br/>
+              cos.flush();<br/>
+              cos.close();<br/>
+              // Decrypt<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: // Decryption with CryptoInputStream

File: src/site/xdoc/userguide.xml [hunk 28]
```diff
@@ -21,12 +21,155 @@
...
+              CryptoInputStream cis = new CryptoInputStream(new ByteArrayInputStream(os.toByteArray()), cipher, bufferSize, key, iv);<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: // Constructs the original InputStream<br/>
InputStream inputStream = new ByteArrayInputStream(os.toByteArray());<br/>
// Constructs the instance of CryptoInputStream<br/>
CryptoInputStream cis = new CryptoInputStream(inputStream, cipher, bufferSize, key, iv);<br/>

File: src/site/xdoc/userguide.xml [hunk 29]
```diff
@@ -21,12 +21,155 @@
...
+              int decryptedLen = cis.read(decryptedData, 0, 1024);<br/>
```
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: @dianfu do we need to refresh the cipher and add close?

File: src/site/xdoc/userguide.xml (hunk 24 above)
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: OutputStream outputStream = new ByteArrayOutputStream()

File: src/site/xdoc/userguide.xml (hunk 26 above)
Comment (sundapeng) [inline src/site/xdoc/userguide.xml]: os -> outputStream

**Comment by sundapeng** [comment]
Thank @JkSelf , I left some comments.

File: src/site/xdoc/userguide.xml (hunk 29 above)
Comment (dianfu) [inline src/site/xdoc/userguide.xml]: Yes, we need to do that.

**Comment by sundapeng** [comment]
Thank @JkSelf for your contribution

**Pull request merged**